src/pipeline/src/etl/ctx_req.rs

```diff
@@ -0,0 +1,143 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::IntoIter;
+use std::collections::BTreeMap;
+use std::sync::Arc;
+
+use ahash::{HashMap, HashMapExt};
+use api::v1::{RowInsertRequest, RowInsertRequests, Rows};
+use itertools::Itertools;
+use session::context::{QueryContext, QueryContextRef};
+use session::hints::{HINTS_KEY_PREFIX, HINT_KEYS};
+
+use crate::PipelineMap;
+
+const DEFAULT_OPT: &str = "";
+
+// Remove hints from the pipeline context and form a option string
+// e.g: skip_wal=true,ttl=1d
+pub fn from_pipeline_map_to_opt(pipeline_map: &mut PipelineMap) -> String {
+    let mut btreemap = BTreeMap::new();
+    for k in HINT_KEYS {
+        if let Some(v) = pipeline_map.remove(k) {
+            btreemap.insert(k, v.to_str_value());
+        }
+    }
+    btreemap
+        .into_iter()
+        .map(|(k, v)| format!("{}={}", k.replace(HINTS_KEY_PREFIX, ""), v))
+        .join(",")
+}
+
+// split the option string back to a map
+fn from_opt_to_map(opt: &str) -> HashMap<&str, &str> {
+    opt.split(',')
+        .filter_map(|s| {
+            s.split_once("=")
+                .filter(|(k, v)| !k.is_empty() && !v.is_empty())
+        })
+        .collect()
+}
+
+// ContextReq is a collection of row insert requests with different options.
+// The default option is empty string.
+// Because options are set in query context, we have to split them into sequential calls
+// e.g:
+// {
+//     "skip_wal=true,ttl=1d": [RowInsertRequest],
+//     "ttl=1d": [RowInsertRequest],
+// }
+#[derive(Debug, Default)]
+pub struct ContextReq {
+    req: HashMap<String, Vec<RowInsertRequest>>,
+}
+
+impl ContextReq {
+    pub fn from_opt_map(opt_map: HashMap<String, Rows>, table_name: String) -> Self {
+        Self {
+            req: opt_map
+                .into_iter()
+                .map(|(opt, rows)| {
+                    (
+                        opt,
+                        vec![RowInsertRequest {
+                            table_name: table_name.clone(),
+                            rows: Some(rows),
+                        }],
+                    )
+                })
+                .collect::<HashMap<String, Vec<RowInsertRequest>>>(),
+        }
+    }
+
+    pub fn default_opt_with_reqs(reqs: Vec<RowInsertRequest>) -> Self {
+        let mut req_map = HashMap::new();
+        req_map.insert(DEFAULT_OPT.to_string(), reqs);
+        Self { req: req_map }
+    }
+
+    pub fn add_rows(&mut self, opt: String, req: RowInsertRequest) {
+        self.req.entry(opt).or_default().push(req);
+    }
+
+    pub fn merge(&mut self, other: Self) {
+        for (opt, req) in other.req {
+            self.req.entry(opt).or_default().extend(req);
+        }
+    }
+
+    pub fn as_req_iter(self, ctx: QueryContextRef) -> ContextReqIter {
+        let ctx = ctx.clone_inner();
```

Comment:
we should be able to use:

```suggestion
        let ctx = (*ctx).clone();
```